### PR TITLE
Raise error for use_local_synchronization=True and ncclCommSplit

### PR DESCRIPTION
**Pull request by nvcastet**
When there is a device bound to the default group, 'ncclCommSplit' will be used instead of 'ncclCommInitRank'.
The API 'ncclCommSplit' requires all the ranks from the parent communicator to participate.
Ranks not part of this sub-group cannot exit early, they need to call 'perform_nocolor_split' inside '_new_process_group_helper'.

Therefore, raise an error when 'use_local_synchronization=True' is set and ncclCommSplit path will be taken.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @tianyu-l @chauhang @eqy

### CODE REVIEW

**Review by eqy — COMMENTED** [review]
Would it be possible to add a test for this?

**Comment by nvcastet** [comment]
> Would it be possible to add a test for this?

Good point! Just pushed a test.

**Comment by nvcastet** [comment]
@shuqiangzhang in case you have time to review.

File: torch/distributed/distributed_c10d.py [hunk 1]
```diff
@@ -4414,6 +4414,21 @@ def _new_group_with_tag(
...
+            # Ranks not part of this sub-group will call 'perform_nocolor_split'.
+            if (
+                is_initialized()
+                and _get_default_group().bound_device_id
+                and len(ranks) < _get_default_group().size()
+            ):
+                raise ValueError(
+                    "NCCL backend doesn't support use_local_synchronization=True "
+                    "in sub-groups when device_id is used with init_process_group"
+                )
```
Comment (kwen2501) [inline torch/distributed/distributed_c10d.py:4431]: Hmm, does that mean the `use_local_synchronization` is less relevant when `device_id` is provided? Maybe in a future refactorization we can entirely separate the two?

File: test/distributed/test_c10d_nccl.py [hunk 2]
```diff
@@ -670,6 +670,19 @@ def test_non_blocking_with_eager_init(self):
...
+    @skip_but_pass_in_sandcastle_if(not TEST_MULTIGPU, "NCCL test requires 2+ GPUs")
+    def test_new_group_device_bound_local_sync_not_allowed(self):
+        store = c10d.FileStore(self.file_name, self.world_size)
+        device = torch.device(f"cuda:{self.rank}")
+        self._create_process_group_nccl(store, self.opts(), device_id=device)
+        with self.assertRaisesRegex(
+            ValueError,
+            "NCCL backend doesn't support use_local_synchronization=True "
+            "in sub-groups when device_id is used with init_process_group",
+        ):
```
Comment (kwen2501) [inline test/distributed/test_c10d_nccl.py:683]: nit: shall we relax the matching condition or just expect a ValueError? That would be easier for a later developer if they happen need to change the error message.

**Review by kwen2501 — APPROVED** [review]
LGTM

**Pull request closed without merging**